napari/_vispy/visuals/volume.py

```diff
@@ -171,6 +95,127 @@
 }
 """
 
+FAST_GRADIENT_DEFINITION = """
+vec3 calculateGradient(vec3 loc, vec3 step) {
+    // calculate gradient within the volume by finite differences
+
+    vec3 G = vec3(0.0);
+
+    float prev;
+    float next;
+    int in_bounds;
+    float curr = colorToVal($get_data(loc));
+
+    for (int i=0; i<3; i++) {
+        vec3 ax_step = vec3(0.0);
+        ax_step[i] = step[i];
+
+        vec3 prev_loc = loc - ax_step;
+        if (u_clamp_at_border || (prev_loc[i] >= 0.0 && prev_loc[i] <= 1.0)) {
+            prev = colorToVal($get_data(prev_loc));
+        } else {
+            prev = categorical_bg_value;
+        }
+
+        vec3 next_loc = loc + ax_step;
+        if (u_clamp_at_border || (next_loc[i] >= 0.0 && next_loc[i] <= 1.0)) {
+            next = colorToVal($get_data(next_loc));
+        } else {
+            next = categorical_bg_value;
+        }
```

Comment:
This is checking whether the sample will be in-bounds for texture sampling. Texture coordinates range from 0 to 1 (inclusive).

If it's in bounds or if we want the default behavior (clamping), we sample the texture. Otherwise if it's out of bounds, assume the background value.

Reply:
This is the bit I was asking about re: performance @aganders3 — the if statement is new (?) and in-loop so could be significantly slower?

btw I don't see the previous fast gradient being removed in this diff — is it because it's now our own shader whereas before it was in VisPy?

Reply:
The if-statement here is indeed new, and is effectively executed 2 x 3 = 6 times per fragment. I don't expect a measurable performance impact here and I think the benchmarks confirm that so far. We can investigate further if you're still worried about it!

See https://github.com/napari/napari/pull/7431/files#r1873897094 for where the previous code is deleted.